sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala

```diff
@@ -232,15 +233,35 @@ case class RelationConversions(
           if DDLUtils.isHiveTable(relation.tableMeta) && isConvertible(relation) =>
         metastoreCatalog.convert(relation, isWrite = false)
 
-      // CTAS
-      case CreateTable(tableDesc, mode, Some(query))
+      // CTAS path
+      // `InsertIntoHiveTable` is derived from `CreateHiveTableAsSelectCommand`.
+      // This pattern would not cause conflicts because this rule is always applied before
+      // `HiveAnalysis` and both of these rules are running once.
+      case InsertIntoHiveTable(tableDesc, _, query, overwrite, ifPartitionNotExists, _)
```

Comment:
wait, how do we optimize hive table insertion before? Given there is no case for `InsertIntoHiveTable` before.

Reply:
Before, we optimize hive table insertion inside `OptimizedCreateHiveTableAsSelectCommand`. It returns `InsertIntoHadoopFsRelationCommand` in `getWritingCommand`. Now, `CreateHiveTableAsSelectCommand` always use `InsertIntoHiveTable` for data writing and we match this pattern to do optimize.

Reply:
I mean normal table insertion, not table insertion inside CTAS.

Reply:
Before we match `InsertIntoStatement` and covert hive relation to HadoopFsRelation then datasource can tune it to `InsertIntoHadoopFsRelationCommand`. see line 223

That said, we can make `CreateHiveTableAsSelectCommand` use `InsertIntoStatement` instead of `InsertIntoHiveTable` but one issue is we can not aware of if the `InsertIntoStatement` is from CTAS. So here I use `InsertIntoHiveTable`.

Reply:
I see, can we make the code comments clearer? This only matches table insertion inside Hive CTAS.

Reply:
updated comment